### PR TITLE
Block Editor: update block-editor routes

### PR DESCRIPTION
**Pull request by retrofox**
~# IMPORTANT: DO NOT MERGE~
~This PR has been branched off from `remove/selected-editor-state`.~

### Changes proposed in this Pull Request

This PR updates the routes for the editor, removing the `/block-editor` string. Also, it updates the links for the post, pages, etc.

Screen Shot 2020-10-13 at 1 53 52 PM

The routes, defined by the not-existing post-editor file on this PR, are the following:

```
page( '/post', controller.pressThis, siteSelection, sites, makeLayout, clientRender );
page( '/post/new', () => page.redirect( '/post' ) ); // redirect from beep-beep-boop
page( '/post/:site?/:post?', siteSelection, controller.gutenberg, controller.post, makeLayout, clientRender );
page.exit( '/post/:site?/:post?', controller.exitPost );

page( '/page', siteSelection, sites, makeLayout, clientRender );
page( '/page/new', () => page.redirect( '/page' ) ); // redirect from beep-beep-boop
page( '/page/:site?/:post?', siteSelection, controller.gutenberg, controller.post, makeLayout, clientRender );
page.exit( '/page/:site?/:post?', controller.exitPost );

page( '/edit/:type', siteSelection, sites, makeLayout, clientRender );
page( '/edit/:type/new', ( context ) => page.redirect( `/edit/${ context.params.type }` ) );
page( '/edit/:type/:site?/:post?', siteSelection, controller.gutenberg, controller.post, makeLayout, clientRender );
page.exit( '/edit/:type/:site?/:post?', controller.exitPost );
```

We need to check that these routes are properly covered after these changes.

* Fixes https://github.com/Automattic/wp-calypso/issues/45515.

#### Before this PR (prod now)

...when a post is edited with the blocks editor, the URL has the following shape:

`https://wordpress.com/create-block/post/<site-id>/<post-id>`

and when is the classic editor the default one, the URL changes to

`https://wordpress.com/post/<site-id>/<post-id>`.

It worths to mention the app redirects to the proper URL depending on what's the current editor: `classic` or `blocks`.

#### After this PR

The app will support only the blocks editor, meaning that now the default URL doesn't have the `/block-editor` string on it. In other words, it preserves the URL shape of the (old) classic editor.

Additionally, we've added a redirection when the URL contains the `/block-editor` string on it, trying to guarantee to get the proper direction for olds URLs.

### Testing instructions

You should test all possibilities/routes when editing content, comparing the behavior with production.

#### Post context
```
http://calypso.localhost:3000/post
http://calypso.localhost:3000/post/<site-id>
http://calypso.localhost:3000/post/<site-id>/new
http://calypso.localhost:3000/post/<site-id>/<post-id>
```

#### Post context - redirection

```
http://calypso.localhost:3000/block-editor
http://calypso.localhost:3000/block-editor/post
http://calypso.localhost:3000/block-editor/post/<site-id>
http://calypso.localhost:3000/block-editor/post/<site-id>/new
http://calypso.localhost:3000/block-editor/post/<site-id>/<post-id>
```

#### Page context

```
http://calypso.localhost:3000/page/
http://calypso.localhost:3000/page/<site-id>
http://calypso.localhost:3000/page/<site-id>/new
http://calypso.localhost:3000/page/<site-id>/<page-id>
```

#### Page context - redirection

```
http://calypso.localhost:3000/block-editor
http://calypso.localhost:3000/block-editor/page
http://calypso.localhost:3000/block-editor/page/<site-id>
http://calypso.localhost:3000/block-editor/page/<site-id>/new
http://calypso.localhost:3000/block-editor/page/<site-id>/<page-id>
```

### CODE REVIEW

File: client/gutenberg/editor/index.js
Comment (retrofox) [inline client/gutenberg/editor/index.js:21]: I'm adding this middleware because it was in the previous post-editor component. I'm no getting totally why we need it, so looking into worths it.

File: client/gutenberg/editor/controller.js
Comment (allilevine) [inline client/gutenberg/editor/controller.js:167]: I was also looking at [removing this code](https://github.com/Automattic/wp-calypso/pull/45938/files#r497586206), but now I'm wondering if it's still needed to [check if the editor flow](https://github.com/Automattic/wp-calypso/blob/master/client/state/selectors/get-selected-editor.js) is forced classic, redirected to wp-admin, or Calypsoified.

Comment (retrofox) [reply client/gutenberg/editor/controller.js:167]: that's a very good point. We need revert removing this function, and add it as middleware of the routes.

Comment (mmtr) [reply client/gutenberg/editor/controller.js:167]: We should also add the `waitForSiteIdAndSelectedEditor` logic back then. Otherwise the selected editor state could be potentially not set when visiting directly a `/post` or `/block-editor` URL (the selected editor state is set [here](https://github.com/Automattic/wp-calypso/blob/c8343b8745b408889b8d1384c4f91f7079671b48/client/layout/index.jsx#L183)).

File: client/gutenberg/editor/controller.js
Comment (retrofox) [inline client/gutenberg/editor/controller.js:196]: I don't get the whole logic totally tbh, but I presume, after these changes, we should just `return next()` here? Since we are not redirecting anymore.

Comment (mmtr) [reply client/gutenberg/editor/controller.js:196]: Right, since `shouldLoadGutenberg` always returns true now, we can remove the `shouldLoadGutenberg` check above and always `return next()`.

**Comment by retrofox** [comment]
Let's hold on until https://github.com/Automattic/wp-calypso/pull/45938 merges.

**Review by wp-desktop — DISMISSED** [review]
WordPress Desktop CI Failure for job "wp-desktop-mac".

@retrofox please inspect this job's build steps for breaking changes at [this link](https://circleci.com/gh/Automattic/wp-calypso/920101). For temporal failures, you may try to "Rerun Workflow from Failed".

Please also ensure this branch is rebased off latest Calypso.

**Comment by kwight** [comment]
I was going to ask if this affects the desktop app at all, but it appears to have verified itself and is satisfied? (I welcome our robot overlords.)

I spot-checked a variety of posts, pages, and redirects, no surprises at all. 🎉

File: client/gutenberg/editor/index.js
Comment (pablinos) [reply client/gutenberg/editor/index.js:36]: I did a bit of research into this, and it essentially clears the `postId` from the state. There is a `startEditingPost` action that occurs in the calypsoify iframe, which does the opposite, and sets the `postId` in the state. This is apparently for components like editor revisions, but I assume that these work outside of Gutenframe, probably via the editing toolkit.

Even if that's not the case, it's not being cleared currently when people leave the Gutenframe editor, so it's probably not essential, but I suppose it's good to keep things tidy.

I'll do a bit more investigation into the editor components using it, so we can make a final decision.

File: client/gutenberg/editor/index.js
Comment (mmtr) [inline client/gutenberg/editor/index.js:86]: I think old URLs for custom types (`/block-editor/edit/:customPostType`) should be included here as well.

Comment (retrofox) [reply client/gutenberg/editor/index.js:86]: addressed here c5c99be7a95b9a55c5bf317dcdcc5bdc0727b314

**Comment by allilevine** [comment]
I tested on calypso.live and all of the routes listed redirected correctly except https://hash-c5c99be7a95b9a55c5bf317dcdcc5bdc0727b314.calypso.live/block-editor/ -- that loads a blank page on this branch but redirects to block-editor/post on staging.

Also, is there an issue to update the default routes on /posts and /pages so that they no longer go to /block-editor?

**Comment by retrofox** [comment]
> I tested on calypso.live and all of the routes listed redirected correctly except https://hash-c5c99be7a95b9a55c5bf317dcdcc5bdc0727b314.calypso.live/block-editor/ -- that loads a blank page on this branch but redirects to block-editor/post on staging.

Looking

> 
> Also, is there an issue to update the default routes on /posts and /pages so that they no longer go to /block-editor?

We need to update these routes, and I thought it could be addressed in another PR?

**Comment by retrofox** [comment]
> I tested on calypso.live and all of the routes listed redirected correctly except https://hash-c5c99be7a95b9a55c5bf317dcdcc5bdc0727b314.calypso.live/block-editor/ -- that loads a blank page on this branch but redirects to block-editor/post on staging.

Addressed [here](https://github.com/Automattic/wp-calypso/pull/46006/commits/068e57b38c1f9ad9ba7aa17803079770dd375a0e)

**Comment by allilevine** [comment]
I tested on calypso.live and all of the routes listed redirected correctly except https://hash-c5c99be7a95b9a55c5bf317dcdcc5bdc0727b314.calypso.live/block-editor/ -- that loads a blank page on this branch but redirects to block-editor/post on staging.

Also, is there an issue to update the default routes on /posts and /pages so that they no longer go to /block-editor?

> > I tested on calypso.live and all of the routes listed redirected correctly except https://hash-c5c99be7a95b9a55c5bf317dcdcc5bdc0727b314.calypso.live/block-editor/ -- that loads a blank page on this branch but redirects to block-editor/post on staging.
> 
> Addressed [here](https://github.com/Automattic/wp-calypso/pull/46006/commits/068e57b38c1f9ad9ba7aa17803079770dd375a0e)

Looks good! It redirects to /post now. 👍

File: client/gutenberg/editor/index.js
Comment (allilevine) [reply client/gutenberg/editor/index.js:87]: > addressed here [c5c99be](https://github.com/Automattic/wp-calypso/commit/c5c99be7a95b9a55c5bf317dcdcc5bdc0727b314)

I tested this on http://calypso.localhost:3000 and when I click to edit a jetpack portfolio (/block-editor/edit/jetpack-portfolio/<site-id>/<page-id>) I get redirected to /edit/jetpack-portfolio/<site-id> to create a new project. 🤔

File: client/gutenberg/editor/index.js
Comment (mmtr) [inline client/gutenberg/editor/index.js:91]: I thinks this is causing the odd behavior noted by @allilevine in https://github.com/Automattic/wp-calypso/pull/46006#discussion_r503567121

```suggestion
			const { customPostType, site, post: postId } = params;
```

Comment (retrofox) [reply client/gutenberg/editor/index.js]: Thanks for this, Miguel.

**Review by allilevine — APPROVED** [review]
I tested all of the edit and create routes for posts, pages, and custom post types and all of them redirect correctly. 🎉

**Review by sirbrillig — APPROVED** [review]
Approved for Shilling team.

File: client/my-sites/sites/index.jsx
Comment (mmtr) [inline client/my-sites/sites/index.jsx:40]: I think we don't want to exclude VIP sites here anymore. This component renders a site selector when accessing routes without a site param:

Screen Shot 2020-10-14 at 11 10 42

We were excluding VIP sites on `/block-editor` routes because there was no support for Gutenberg for them, but site selectors rendered on `/post` or `/page` routes should include VIP sites because that's a working route for them (I guess end up using the classic editor).

Comment (retrofox) [reply client/my-sites/sites/index.jsx:40]: makes sense

**Review by mmtr — APPROVED** [review]

**Review by creativecoder — CHANGES_REQUESTED** [review]
I've having an issue with an Atomic test site: it lands on /wp-admin when creating a new post (or editing an existing one) from Calypso. This is testing with calypso.live

Atomic - New post - redirect to wp-admin

**Comment by pablinos** [comment]
That looks like the issue solved by #46351. You shouldn't have a problem testing with calypso.localhost

**Review by creativecoder — APPROVED** [review]
I tried loading the editor in Calypso for various sites (Simple and Atomic), post types (post, page, jetpack-portfolio), and redirects for `/block-editor`. All working as expected.

**Comment by creativecoder** [comment]
I'm not sure if the e2e test failures are related or not, but we should figure that out before merging.

**Pull request merged**
